From 5c26e9881db6856306b6fb5a74f92b494a8c848a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Cl=C3=A9ment=20Le=20Goffic?= <clement.legoffic@foss.st.com>
Date: Fri, 29 Nov 2024 15:03:40 +0100
Subject: [PATCH] bus: stm32: move stm32_firewall_populate_bus() function
 inside drivers
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Firewall controller drivers can chose to apply different firewall
configuration on probe of bus children.

Change-Id: Id10af3df80fd03febedeac87b1259e7fb965ab45
Signed-off-by: Clément Le Goffic <clement.legoffic@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/421712
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/bus/stm32_etzpc.c          | 57 ++++++++++++++++++++++++++++-
 drivers/bus/stm32_firewall.c       | 55 ----------------------------
 drivers/bus/stm32_rifsc.c          | 58 +++++++++++++++++++++++++++++-
 include/linux/bus/stm32_firewall.h | 12 -------
 4 files changed, 113 insertions(+), 69 deletions(-)

--- a/drivers/bus/stm32_etzpc.c
+++ b/drivers/bus/stm32_etzpc.c
@@ -6,6 +6,7 @@
 #include <linux/bitfield.h>
 #include <linux/bits.h>
 #include <linux/bus/stm32_firewall.h>
+#include <linux/bus/stm32_firewall_device.h>
 #include <linux/device.h>
 #include <linux/err.h>
 #include <linux/init.h>
@@ -15,6 +16,7 @@
 #include <linux/of.h>
 #include <linux/of_platform.h>
 #include <linux/platform_device.h>
+#include <linux/slab.h>
 #include <linux/types.h>
 
 /*
@@ -69,6 +71,59 @@ static void stm32_etzpc_release_access(s
 {
 }
 
+static int stm32_etzpc_populate_bus(struct stm32_firewall_controller *ctrl)
+{
+	struct stm32_firewall *firewalls;
+	struct device_node *child;
+	struct device *parent;
+	unsigned int i;
+	int len;
+	int err;
+
+	parent = ctrl->dev;
+
+	dev_dbg(parent, "Populating %s system bus\n", dev_name(ctrl->dev));
+
+	for_each_available_child_of_node(dev_of_node(parent), child) {
+		/* The access-controllers property is mandatory for firewall bus devices */
+		len = of_count_phandle_with_args(child, "access-controllers",
+						 "#access-controller-cells");
+		if (len <= 0) {
+			of_node_put(child);
+			return -EINVAL;
+		}
+
+		firewalls = kcalloc(len, sizeof(*firewalls), GFP_KERNEL);
+		if (!firewalls) {
+			of_node_put(child);
+			return -ENOMEM;
+		}
+
+		err = stm32_firewall_get_firewall(child, firewalls, (unsigned int)len);
+		if (err) {
+			kfree(firewalls);
+			of_node_put(child);
+			return err;
+		}
+
+		for (i = 0; i < len; i++) {
+			if (ctrl->grant_access(ctrl, firewalls[i].firewall_id)) {
+				/*
+				 * Peripheral access not allowed or not defined.
+				 * Mark the node as populated so platform bus won't probe it
+				 */
+				of_detach_node(child);
+				dev_err(parent, "%s: Device driver will not be probed\n",
+					child->full_name);
+			}
+		}
+
+		kfree(firewalls);
+	}
+
+	return 0;
+}
+
 static int stm32_etzpc_probe(struct platform_device *pdev)
 {
 	struct stm32_firewall_controller *etzpc_controller;
@@ -109,7 +164,7 @@ static int stm32_etzpc_probe(struct plat
 		return rc;
 	}
 
-	rc = stm32_firewall_populate_bus(etzpc_controller);
+	rc = stm32_etzpc_populate_bus(etzpc_controller);
 	if (rc) {
 		dev_err(etzpc_controller->dev, "Couldn't populate ETZPC bus: %d",
 			rc);
--- a/drivers/bus/stm32_firewall.c
+++ b/drivers/bus/stm32_firewall.c
@@ -236,58 +236,3 @@ void stm32_firewall_controller_unregiste
 			 firewall_controller->name);
 }
 EXPORT_SYMBOL_GPL(stm32_firewall_controller_unregister);
-
-int stm32_firewall_populate_bus(struct stm32_firewall_controller *firewall_controller)
-{
-	struct stm32_firewall *firewalls;
-	struct device_node *child;
-	struct device *parent;
-	unsigned int i;
-	int len;
-	int err;
-
-	parent = firewall_controller->dev;
-
-	dev_dbg(parent, "Populating %s system bus\n", dev_name(firewall_controller->dev));
-
-	for_each_available_child_of_node(dev_of_node(parent), child) {
-		/* The access-controllers property is mandatory for firewall bus devices */
-		len = of_count_phandle_with_args(child, "access-controllers",
-						 "#access-controller-cells");
-		if (len <= 0) {
-			of_node_put(child);
-			return -EINVAL;
-		}
-
-		firewalls = kcalloc(len, sizeof(*firewalls), GFP_KERNEL);
-		if (!firewalls) {
-			of_node_put(child);
-			return -ENOMEM;
-		}
-
-		err = stm32_firewall_get_firewall(child, firewalls, (unsigned int)len);
-		if (err) {
-			kfree(firewalls);
-			of_node_put(child);
-			return err;
-		}
-
-		for (i = 0; i < len; i++) {
-			if (firewall_controller->grant_access(firewall_controller,
-							      firewalls[i].firewall_id)) {
-				/*
-				 * Peripheral access not allowed or not defined.
-				 * Mark the node as populated so platform bus won't probe it
-				 */
-				of_detach_node(child);
-				dev_err(parent, "%s: Device driver will not be probed\n",
-					child->full_name);
-			}
-		}
-
-		kfree(firewalls);
-	}
-
-	return 0;
-}
-EXPORT_SYMBOL_GPL(stm32_firewall_populate_bus);
--- a/drivers/bus/stm32_rifsc.c
+++ b/drivers/bus/stm32_rifsc.c
@@ -18,6 +18,7 @@
 #include <linux/types.h>
 
 #include <linux/bus/stm32_firewall.h>
+#include <linux/bus/stm32_firewall_device.h>
 
 /*
  * RIFSC offset register
@@ -575,6 +576,61 @@ static void stm32_rifsc_release_access(s
 	stm32_rif_release_semaphore(ctrl, firewall_id);
 }
 
+static int stm32_rifsc_populate_bus(struct stm32_firewall_controller *ctrl)
+{
+	struct stm32_firewall *firewalls;
+	struct device_node *child;
+	struct device *parent;
+	unsigned int i;
+	int len;
+	int err;
+
+	parent = ctrl->dev;
+
+	dev_dbg(parent, "Populating %s system bus\n", dev_name(ctrl->dev));
+
+	for_each_available_child_of_node(dev_of_node(parent), child) {
+		/* The access-controllers property is mandatory for firewall bus devices */
+		len = of_count_phandle_with_args(child, "access-controllers",
+						 "#access-controller-cells");
+		if (len <= 0) {
+			of_node_put(child);
+			return -EINVAL;
+		}
+
+		firewalls = kcalloc(len, sizeof(*firewalls), GFP_KERNEL);
+		if (!firewalls) {
+			of_node_put(child);
+			return -ENOMEM;
+		}
+
+		err = stm32_firewall_get_firewall(child, firewalls, (unsigned int)len);
+		if (err) {
+			kfree(firewalls);
+			of_node_put(child);
+			return err;
+		}
+
+		for (i = 0; i < len; i++) {
+			err = firewalls[i].firewall_ctrl->grant_access(firewalls[i].firewall_ctrl,
+								       firewalls[i].firewall_id);
+			if (err) {
+				/*
+				 * Peripheral access not allowed or not defined.
+				 * Mark the node as populated so platform bus won't probe it
+				 */
+				of_detach_node(child);
+				dev_err(parent, "%s: Device driver will not be probed, error: %d\n",
+					child->full_name, err);
+			}
+		}
+
+		kfree(firewalls);
+	}
+
+	return 0;
+}
+
 static int stm32_rifsc_probe(struct platform_device *pdev)
 {
 	struct stm32_firewall_controller *rifsc_controller;
@@ -657,7 +713,7 @@ static int stm32_rifsc_probe(struct plat
 		return rc;
 	}
 
-	rc = stm32_firewall_populate_bus(rifsc_controller);
+	rc = stm32_rifsc_populate_bus(rifsc_controller);
 	if (rc) {
 		dev_err(rifsc_controller->dev, "Couldn't populate RIFSC bus: %d",
 			rc);
--- a/include/linux/bus/stm32_firewall.h
+++ b/include/linux/bus/stm32_firewall.h
@@ -70,16 +70,4 @@ int stm32_firewall_controller_register(s
  */
 void stm32_firewall_controller_unregister(struct stm32_firewall_controller *firewall_controller);
 
-/**
- * stm32_firewall_populate_bus - Populate device tree nodes that have a correct firewall
- *				 configuration. This is used at boot-time only, as a sanity check
- *				 between device tree and firewalls hardware configurations to
- *				 prevent a kernel crash when a device driver is not granted access
- *
- * @firewall_controller:	Firewall controller which nodes will be populated or not
- *
- * Returns 0 in case of success or appropriate errno code if error occurred.
- */
-int stm32_firewall_populate_bus(struct stm32_firewall_controller *firewall_controller);
-
 #endif /* _STM32_FIREWALL_H */
